File: Cpp/LeetCode424.LongestRepeatingCharacterReplacement_String.cpp
// REFERENCE: https://www.youtube.com/watch?v=gqXU1UyA8pk

#include <unordered_map>
#include <string>

class Solution {
public:
    int characterReplacement(std::string s, int k) {
        std::unordered_map<char, int> countmap;
        int l = 0;
        int r = 0;
        int maxF = 0;
        int ans = 0;
        while (r<s.length())
        {
            if (countmap[s[r] <= 0])
            {
                countmap[s[r]] = 1;
            } else {
                countmap[s[r]]++;
            }
            
            maxF = std::max(maxF, countmap[s[r]]);
            
            while ((r-l+1) - maxF > k)
            {
                countmap[s[l]]--;
                l++;
            }
            ans = std::max(ans, r-l+1);
            r++;
        }
        return ans;
    }
};

File: Python/LeetCode424.LongestRepeatingCharacterReplacement_String.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        l = 0
        r = 0
        max_frequency = 0
        max_length = 0
        char_count = {}

        while r < len(s):
            char_count[s[r]] = char_count.get(s[r], 0) + 1
            max_frequency = max(max_frequency, char_count[s[r]])

            while r-l+1 - max_frequency > k:
                char_count[s[l]] -= 1
                l += 1

            max_length = max(max_length, r-l+1)
            r += 1

        return max_length
